File: top100liked/fastandslowpointers/LinkedListCycle.java
package fastandslowpointers;

import java.util.*;

public class LinkedListCycle {

    public static void main(String[] args) {
        System.out.println(hasCycle(
                new ListNode(3, new ListNode(2, new ListNode(0, new ListNode(-4, new ListNode(1)))))
        ));
    }

    public static boolean hasCycle(ListNode head) {
        ListNode slow = head, fast = head;

        while(fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
            if (fast == slow) {
                return true;
            }
        }

        return false;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
        next = null;
    }

    ListNode(int x, ListNode next) {
        this.val = x;
        this.next = next;
    }

    @Override
    public String toString() {
        return "ListNode{" +
                "val=" + val +
                ", next=" + next +
                '}';
    }
}

File: python/week1/LinkedListCycle.py
from typing import Optional

from ListNode import ListNode


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return False

        slow, fast = head, head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                return True

        return False

if __name__ == '__main__':
    Solution().hasCycle(ListNode(3, ListNode(2, ListNode(0, ListNode(4)))))
    print()